### ISSUE TITLE
[QUESTION] Should log level be included in HostData?

### ISSUE THREAD

**Issue by jordan-rash** [wasmCloud/weld#140]
For the same reason that StructuredLogging is included...the provider knows what level to log at.

https://github.com/wasmCloud/weld/blob/c9796583046e471bf866e5a081afaf27ac1b73f4/rpc-rs/src/wasmbus_core.rs#L328-L330

Or am I missing it somewhere?

**Comment by connorsmith256** [comment]
I had to double-check how this works today, since `RUST_LOG` isn't found anywhere in the OTP host codebase.

Today, provider processes inherit environment variables from their parent, the host process. So if you start the host with `RUST_LOG=debug`, every Rust provider instance started on that host will also have that environment variable set when starting, so they will log at the debug level.

I'm guessing you're looking for an equivalent for a Go provider? 😉 I'm not sure if there's a standard env var like there is for Rust, but the same process would work.

However, to answer your question: yes, I do think log level should be explicitly passed to the provider rather than implicitly via a language-specific env var. I've been interested in writing an RFC around provider configuration more generally. I'm particularly interested in allowing runtime changes to logging config for providers/actors!

**Comment by jordan-rash** [comment]
Yea, I noticed that some things where using RUST_LOG while others were using, well nothing.  I think the `WASMCLOUD_STRUCTURED_LOG_LEVEL` is a good place to start, but maybe with the "STRUCTURED_LOG" part (so it applies to json or text based logs.

I could just have the go sdks look for `RUST_LOG` env, but i'm afraid when I run for Gopher President in the future, my opposition will find that and my dreams will be shattered.

**Comment by jordan-rash** [comment]
@connorsmith256 I remember when I was creating the `wazero` host that the hostdata has this `EnvValues` item https://github.com/wasmCloud/weld/blob/c9796583046e471bf866e5a081afaf27ac1b73f4/rpc-rs/src/wasmbus_core.rs#L314

I assume that would have things like `RUST_LOG` as well as `WASMCLOUD_STRUCTRED_LOG_LEVEL` in it.  I can try and confirm that later....so my question now is, what determines which variables get pulled into the host_data struct and what just rides that env_values map?

**Comment by connorsmith256** [comment]
The `HostData` struct is defined in smithy here: https://github.com/wasmCloud/interfaces/blob/0fe4ba01cccbe2342a987f01586efa8599557458/core/wasmcloud-core.smithy#L149

I thought all env vars were passed to the provider, but now it looks like `env_values` is empty 🤔 https://github.com/wasmCloud/wasmcloud-otp/blob/47cef826f83ae90e36a7fbbd4c7290a1f4f2d56e/host_core/lib/host_core/vhost/virtual_host.ex/#L306

@autodidaddict can you shine some light here?

**Comment by autodidaddict** [comment]
No implicit env vars are passed for security reasons. That empty list was there in case we had specific wildcards or things that we wanted to forward

**Comment by connorsmith256** [comment]
To answer Jordan's original question, how do providers get access to the `RUST_LOG` env var (assuming they do?)

**Comment by jordan-rash** [comment]
Not specifically `RUST_LOG` as much as `WASMCLOUD_STRUCTURED_LOG_LEVEL`

**Comment by brooksmtownsend** [comment]
So providers can access `RUST_LOG` directly from the environment today, but we could also consider giving an explicit log level so that it's not implicitly just available.

**Comment by connorsmith256** [comment]
@brooksmtownsend Now I'm confused more... I found `propagated_env_vars` in the OTP host, but `RUST_LOG` isn't in there. How are variables like `RUST_LOG`, `WASMCLOUD_STRUCTURED_LOG_LEVEL`, and `WASMCLOUD_STRUCTURED_LOGGING_ENABLED` making it to the provider binary?

**Comment by jordan-rash** [comment]
`WASMCLOUD_STRUCTURED_LOGGING_ENABLED` is in HostData....the others aren't.  

I can't read the Elixr code, but whereever its starting a provider, it should be handing it a JSON of the HostData as arg[1]

**Comment by connorsmith256** [comment]
`HostData` has a `structured_logging` [field](https://github.com/wasmcloud/weld/blob/6c766b23644c216c331b66a1784cffc831816525/rpc-rs/src/wasmbus_core.rs/#L330)

It appears the host is passing a `enable_structured_logging` [field](https://github.com/wasmCloud/wasmcloud-otp/blob/26401ca51b3333ee1b50791d025ad970c81ce5be/host_core/lib/host_core/vhost/virtual_host.ex#L316)

`HostData` is passed via stdin to the process [here](https://github.com/wasmCloud/wasmcloud-otp/blob/26401ca51b3333ee1b50791d025ad970c81ce5be/host_core/lib/host_core/providers/provider_module.ex#L170)

As for env vars, it appears only `propagated_env_vars` are [included](https://github.com/wasmCloud/wasmcloud-otp/blob/26401ca51b3333ee1b50791d025ad970c81ce5be/host_core/lib/host_core/providers/provider_module.ex#L168)

So, my confusion:
- are providers correctly receiving the structured logging field from the host data? If so, how, since the field names don't appear to line up?
- where are env vars like `RUST_LOG` getting passed to the provider process?

**Comment by jordan-rash** [comment]
I can speak for Go Providers, they receive the HostData, from the Host, via standard in [here](https://github.com/wasmCloud/provider-sdk-go/blob/14cfe1cb919cad4b92f2bb5d4c93c93c64a6da39/provider.go#L56).  Then it is parsed into the core.HostData

It is my understanding that the only place a provider provided initialization data, is that stdin process you noted above (for all providers)

So I would assume, in order to add a logging level, we would need to add a LogLevel field to HostData for the providers to ingest when they start

**Comment by connorsmith256** [comment]
@jordan-rash what you said makes sense, but it contradicts my observations, which are when starting a host with `RUST_LOG=debug wash up`, rust providers started on the host do indeed log at the debug level. I'm just very confused as to _how_ that's working today 😅 

@brooksmtownsend @autodidaddict ?

**Comment by jordan-rash** [comment]
Yea, because, correct me if I wrong, they are all being ran in the session you just started wash in.  So by running wash up with that ENV, all child pids, which include the started providers, are running in a session with `RUST_LOG=debug` set.  At that point, does not every invocation of `debug!()` fire?

**Comment by connorsmith256** [comment]
Maybe? The [Port docs](https://hexdocs.pm/elixir/1.14.3/Port.html) don't mention how child processes inherit environment variables. Under the hood it looks like elixir is calling [:erlang.open_port](https://www.erlang.org/doc/man/erlang.html#open_port-2). The docs there say about `env`:

> The environment of the started process is extended using the environment specifications in Env.

I'm not sure whether to infer that "extended" means the parent process is passing all of its env vars to the child process. If that's the case, is it a security concern like @autodidaddict alluded to [here](https://github.com/wasmCloud/weld/issues/140#issuecomment-1435107615)?

**Comment by brooksmtownsend** [comment]
Yeah, that's correct @jordan-rash. The Rust providers get the env var `RUST_LOG` from the session you start wash in, so they kind of _implicitly_ figure out their own log level. I think we should hand down log level in the HostData so that regardless of provider language we know where to find it (RUST_LOG means nothing to Go providers, for example.)

I am concerned with passing the full set of environment variables and didn't know we still did that

**Comment by jordan-rash** [comment]
> I am concerned with passing the full set of environment variables and didn't know we still did that

Per @autodidaddict comment, I dont think the host does...the field has always been blank when I checked it.

**Comment by jordan-rash** [comment]
Not sure why closed this if the 558 PR says 
> This doesn't fix https://github.com/wasmCloud/weld/issues/140, but it is related
This seems to still be an issue?

**Comment by connorsmith256** [comment]
I determined the answer to your question is "yes" :)

The issue for tracking that work is here: https://github.com/wasmCloud/weld/issues/145

**Comment by jordan-rash** [comment]
Thanks Connor :-)
